hudi-spark-datasource/hudi-spark3.5.x/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieSpark35Analysis.scala

```diff
@@ -64,3 +72,132 @@ case class HoodieSpark35DataSourceV2ToV1Fallback(sparkSession: SparkSession) ext
     LogicalRelation(relation, output, catalogTable, isStreaming = false)
   }
 }
+
+/**
+ * In Spark 3.5, the following Resolution rules are removed,
+ * [[ResolveUserSpecifiedColumns]] and [[ResolveDefaultColumns]]
+ * (see code changes in [[org.apache.spark.sql.catalyst.analysis.Analyzer]]
+ * from https://github.com/apache/spark/pull/41262).
+ * The same logic of resolving the user specified columns and default values,
+ * which are required for a subset of columns as user specified compared to the table
+ * schema to work properly, are deferred to [[PreprocessTableInsertion]] for v1 INSERT.
+ *
+ * Note that [[HoodieAnalysis]] intercepts the [[InsertIntoStatement]] after Spark's built-in
+ * Resolution rules are applies, the logic of resolving the user specified columns and default
+ * values may no longer be applied. To make INSERT with a subset of columns specified by user
+ * to work, this custom resolution rule [[HoodieSpark35ResolveColumnsForInsertInto]] is added
+ * to achieve the same, before converting [[InsertIntoStatement]] into
+ * [[InsertIntoHoodieTableCommand]].
+ *
+ * Also note that, the project logic in [[ResolveImplementationsEarly]] for INSERT is still
+ * needed in the case of INSERT with all columns in a different ordering.
+ */
+case class HoodieSpark35ResolveColumnsForInsertInto() extends ResolveInsertionBase {
+  // NOTE: This is copied from [[PreprocessTableInsertion]] with additional handling of Hudi relations
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      case i@InsertIntoStatement(table, _, _, query, _, _, _)
+        if table.resolved && query.resolved
+          && i.userSpecifiedCols.nonEmpty && i.table.isInstanceOf[LogicalRelation]
+          && sparkAdapter.isHoodieTable(i.table.asInstanceOf[LogicalRelation].catalogTable.get) =>
+        table match {
+          case relation: HiveTableRelation =>
+            val metadata = relation.tableMeta
+            preprocess(i, metadata.identifier.quotedString, metadata.partitionSchema,
+              Some(metadata))
+          case LogicalRelation(h: HadoopFsRelation, _, catalogTable, _) =>
+            val tblName = catalogTable.map(_.identifier.quotedString).getOrElse("unknown")
+            preprocess(i, tblName, h.partitionSchema, catalogTable)
+          case LogicalRelation(_: InsertableRelation, _, catalogTable, _) =>
+            preprocess(i, catalogTable)
+          // The two conditions below are adapted to Hudi relations
+          case LogicalRelation(_: EmptyRelation, _, catalogTable, _) =>
+            preprocess(i, catalogTable)
+          case LogicalRelation(_: MergeOnReadSnapshotRelation, _, catalogTable, _) =>
```

Comment:
would it be better to do HoodieBaseRelation?

Reply:
Yes, that's general.  Fixed.